File: 2024/day12.go
package main

import (
	"lukechampine.com/advent/utils"
)

var input = utils.Input(2024, 12)

func main() {
	grid := utils.ToByteGrid(utils.Lines(input))
	g := utils.Grid{X: len(grid[0]) - 1, Y: len(grid) - 1}
	seen := make(map[utils.Pos]bool)
	var sum int
	g.ForEach(func(p utils.Pos) {
		if seen[p] {
			return
		}
		c := grid[p.Y][p.X]
		var area, perimeter int
		utils.Flood(p, func(p utils.Pos) (next []utils.Pos) {
			seen[p] = true
			for _, adj := range p.ValidMoves(g) {
				if grid[adj.Y][adj.X] == c {
					next = append(next, adj)
				}
			}
			area++
			perimeter += 4 - len(next)
			return
		})
		sum += area * perimeter
	})
	utils.Println(sum)

	seen = make(map[utils.Pos]bool)
	sum = 0
	g.ForEach(func(p utils.Pos) {
		if seen[p] {
			return
		}
		c := grid[p.Y][p.X]
		var plot []utils.Pos
		utils.Flood(p, func(p utils.Pos) (next []utils.Pos) {
			seen[p] = true
			plot = append(plot, p)
			for _, adj := range p.ValidMoves(g) {
				if grid[adj.Y][adj.X] == c {
					next = append(next, adj)
				}
			}
			return
		})

		min, max := utils.BoundingBox(plot)
		inPlot := func(x, y int) bool {
			return utils.Any(len(plot), func(i int) bool { return plot[i] == utils.Pos{X: x, Y: y} })
		}
		sides := 0
		for x := min.X; x <= max.X; x++ {
			// edges to the left
			for y := min.Y; y <= max.Y; y++ {
				if inPlot(x, y) && !inPlot(x-1, y) {
					sides++
					for inPlot(x, y) && !inPlot(x-1, y) {
						y++
					}
				}
			}
			// edges to the right
			for y := min.Y; y <= max.Y; y++ {
				if inPlot(x, y) && !inPlot(x+1, y) {
					sides++
					for inPlot(x, y) && !inPlot(x+1, y) {
						y++
					}
				}
			}
		}
		for y := min.Y; y <= max.Y; y++ {
			// edges above
			for x := min.X; x <= max.X; x++ {
				if inPlot(x, y) && !inPlot(x, y-1) {
					sides++
					for inPlot(x, y) && !inPlot(x, y-1) {
						x++
					}
				}
			}
			// edges below
			for x := min.X; x <= max.X; x++ {
				if inPlot(x, y) && !inPlot(x, y+1) {
					sides++
					for inPlot(x, y) && !inPlot(x, y+1) {
						x++
					}
				}
			}
		}

		sum += len(plot) * sides
	})
	utils.Println(sum)
}

File: 2024/rust/src/day12.rs
use std::collections::HashSet;

use crate::utils;

pub fn part1(input: &str) -> String {
    let g = utils::Grid::from_string(input);
    let mut seen = HashSet::new();
    g.iter()
        .map(|(p, c)| {
            if !seen.insert(p) {
                return 0;
            }
            let mut area = 0;
            let mut perimeter = 0;
            utils::flood(
                p,
                |p| g.in_bounds(p) && g[p] == c,
                |p| {
                    g.adj(p)
                        .filter_map(|(a, _)| (g[a] == c).then_some(a))
                        .collect()
                },
            )
            .for_each(|p| {
                seen.insert(p);
                area += 1;
                perimeter += 4 - g.adj(p).filter(|&(a, _)| (g[a] == c)).count();
            });
            area * perimeter
        })
        .sum::<usize>()
        .to_string()
}

fn count_sides(shape: Vec<utils::Point>) -> usize {
    let mut sides = 0;
    let mut seen = HashSet::new();
    for d in vec![
        ((-1, 0).into(), (0, 1).into(), (0, -1).into()),
        ((1, 0).into(), (0, 1).into(), (0, -1).into()),
        ((0, -1).into(), (1, 0).into(), (-1, 0).into()),
        ((0, 1).into(), (1, 0).into(), (-1, 0).into()),
    ] {
        for &p in &shape {
            if seen.contains(&p) || shape.contains(&p.add(d.0)) {
                continue;
            }
            sides += 1;
            seen.extend(utils::flood(
                p,
                |p| shape.contains(&p) && !shape.contains(&p.add(d.0)),
                |p| vec![p.add(d.1), p.add(d.2)],
            ));
        }
        seen.clear();
    }
    sides
}

pub fn part2(input: &str) -> String {
    let g = utils::Grid::from_string(input);
    let mut seen = HashSet::new();
    g.iter()
        .map(|(p, c)| {
            if !seen.insert(p) {
                return 0;
            }
            let shape: Vec<_> = utils::flood(
                p,
                |p| g.in_bounds(p) && g[p] == c,
                |p| {
                    g.adj(p)
                        .filter_map(|(a, _)| (g[a] == c).then_some(a))
                        .collect()
                },
            )
            .collect();
            shape.iter().for_each(|&p| {
                seen.insert(p);
            });

            shape.len() * count_sides(shape)
        })
        .sum::<usize>()
        .to_string()
}
